src/compare-pdf-to-snapshot.ts

```diff
@@ -100,56 +100,87 @@ export type CompareOptions = {
+ *   - If `allowSnapshotCreation` is `true` (default), the PDF is converted to an image,
+ *     saved as a snapshot, and the function returns `true`.
+ *   - If `allowSnapshotCreation` is `false` and `failOnMissingSnapshot` is `true`,
+ *     the function returns `false` without generating a new snapshot.
+ *   - If `allowSnapshotCreation` is `false` and `failOnMissingSnapshot` is `false`
+ *     (default), the function returns `true` without generating a new snapshot.
+ * - If a snapshot exists, the PDF is converted to an image and compared to the snapshot:
+ *   - If they differ, the function returns `false` and creates two additional images
+ *     next to the snapshot: one with the suffix `new` (the current view of the PDF as an image)
+ *     and one with the suffix `diff` (showing the difference between the snapshot and the `new` image).
+ *   - If they are equal, the function returns `true`. If `new` and `diff` versions are present, they are deleted.
+ *
+ * @param pdf - Path to the PDF file or a Buffer containing the PDF.
+ * @param snapshotDir - Path to the directory where the `__snapshots__` folder will be created.
+ * @param snapshotName - Unique name for the snapshot within the specified path.
+ * @param options - Options for comparison, including tolerance, mask regions, and behavior
+ * regarding missing snapshots. See {@link CompareOptions} for more details.
  *
  * @returns
- * A promise that resolves to `true` if the PDF matches the snapshot or
- * if a new snapshot is created, and `false` if the PDF differs from the snapshot.
+ * A promise that resolves to `true` if the PDF matches the snapshot or if the behavior
+ * for a missing snapshot is configured to allow it. Returns `false` if the PDF differs
+ * from the snapshot or if `failOnMissingSnapshot` is `true` and the snapshot is missing.
  */
 export function comparePdfToSnapshot(
-  /** Path to the PDF file or a Buffer containing the PDF. */
   pdf: string | Buffer,
-  /** Path to the directory where `__snapshots__` folder will be created. */
   snapshotDir: string,
-  /** Unique name for the snapshot within the specified path. */
   snapshotName: string,
-  /** Check the type for available options. */
   options?: CompareOptions,
 ): Promise<boolean> {
-  const { maskRegions = () => [], pdf2PngOptions, ...restOpts } = options || {}
+  const {
+    maskRegions = () => [],
+    pdf2PngOptions,
+    allowSnapshotCreation = true,
+    failOnMissingSnapshot = false,
+    ...restOpts
+  } = options || {}
   const dir = path.join(snapshotDir, snapshotsDirName)
   if (!existsSync(dir)) {
     mkdirSync(dir, { recursive: true })
   }
 
   const snapshotPath = path.join(dir, snapshotName + '.png')
 
+  // If the snapshot doesn't exist
   if (!existsSync(snapshotPath)) {
+    // If we shouldn't generate a snapshot, handle based on failIfSnapshotMissing
+    if (!allowSnapshotCreation) {
+      return Promise.resolve(!failOnMissingSnapshot)
```

Comment:
is the negation here intentional? Probably just me being an annoying linter, this might be a bit clearer regarding intention :)

```ts
const snapshotExists = existsSync(snapshotPath);
const failTestDueToMissingSnapshot = (!snapshotExists && failOnMissingSnapshot);
const shouldGenerateSnapshot = (!snapshotExists && allowSnapshotCreation);

if (shouldGenerateSnapshot) {
    return pdf2png(pdf, pdf2PngOptions)
      .then(maskImgWithRegions(maskRegions))
      .then(writeImages(snapshotPath))
      .then(() => failTestDueToMissingSnapshot)
}

return pdf2png(...);
```

I think that achieves the same but it's untested - thanks for the continued work on this project, tag me any time for reviews, happy to contribute when/where needed too.

Reply:
Won't this change the semantics a bit?

> When the function is executed, it has the following **side effects**:
>
> - If a previous snapshot file does not exist:
>   - If `allowSnapshotCreation` is `true` (default), the PDF is converted to an image,
>     saved as a snapshot, and the function returns `true`.
>   - If `allowSnapshotCreation` is `false` and `failOnMissingSnapshot` is `true`,
>     the function returns `false` without generating a new snapshot.
>   - If `allowSnapshotCreation` is `false` and `failOnMissingSnapshot` is `false`
>     (default), the function returns `true` without generating a new snapshot.
> - If a snapshot exists...

This means that `failOnMissingSnapshot` only applies when `allowSnapshotCreation = false`, which clearly deviates from your original intention of keeping these two options independent.

If they are supposed to be independent, what should happen when a snapshot **does not exist** and both `allowSnapshotCreation = true` and `failOnMissingSnapshot = true`?

Let's assume `comparePdfToSnapshot` behaves as follows when no snapshot is present:

| # | `allowSnapshotCreation` | `failOnMissingSnapshot` | Behavior |
| --- | --- | --- | --- |
| 1 | true | true | Creates snapshot and returns `false` |
| 2 | true | false | Creates snapshot and returns `true` |
| 3 | false | true | Does **not** create a snapshot and returns `false` |
| 4 | false | false | Does **not** create a snapshot and returns `true` |

In the first case `#1`, on the next run this code path won't be triggered, because a snapshot will exist, and `comparePdfToSnapshot` will return `true` instead of `false` as it did in the previous run.

In the second case `#2`, the code path also won't be triggered on the next run due to the previously created snapshot, but at least the function will return the same result (`true`).

I prefer consecutive runs to always return the same result (excluding side effects) unless the PDF has changed in the meantime. That’s why the current implementation behaves as it does.

Ultimately, what problem are we solving here? I see a clear case for having a "CI" option that returns `false` when a snapshot doesn't exist. This could ensure that, during a CI test run, any new test also has an associated snapshot committed. If this is the only use case, we could simplify the options significantly.

Reply:
@GrayedFox, @mildfuzz,

Can I have your input?

Reply:
@GrayedFox, thank you for the code improvements. I'd like to postpone them and discuss them as a follow-up to this PR. I prefer to have a separate refactoring PR without any embedded API changes. 

Regarding the API itself, I become more convinced that we should add only one new option: `failOnMissingSnapshot` which would be optional and set to `false` by default. This will enable the CI use case where `comparePdfToSnapshot` returns `false` when the reference snapshot does not exist. 

@mildfuzz, will that suit your use case? 

Having two boolean variables, with potential four different states - where one is ignored by the code and the other might not be used at all - makes me want to simplify things. You can [see here](https://github.com/moshensky/pdf-visual-diff/pull/74) what I mean.

Reply:
The fail on missing snapshot meets our immediate needs for implementing this in our CI, if @mildfuzz feels the same, happy to see just that functionality merged in the other PR and we can think a bit deeper about adding the second option in a way that keeps everything intuitive and aligns with the core design goals and principles of your module

Reply:
The fail on missing snapshot meets our immediate needs for implementing this in our CI, if @mildfuzz feels the same, happy to see just that functionality merged in the other PR and we can think a bit deeper about adding the second option in a way that keeps everything intuitive and aligns with the core design goals and principles of your module

Reply:
I'll move forward with merging just `failOnMissingSnapshot` functionality.
We can revisit adding any additional options later.